pkg/controllers/config/configsync.go

```diff
@@ -0,0 +1,220 @@
+	eventRecorder        events.Recorder
+}
+
+const (
+	sourceConfigKey = "config"
+	targetConfigKey = "cloud.conf"
+
+	infrastructureResourceName = "cluster"
+)
+
+func NewConfigSyncController(
+	operatorClient v1helpers.OperatorClient,
+	kubeClient kubernetes.Interface,
+	informers v1helpers.KubeInformersForNamespaces,
+	configInformers configinformers.SharedInformerFactory,
+	resyncInterval time.Duration,
+	eventRecorder events.Recorder) factory.Controller {
+
+	// Read configmap from user-managed namespace and save the translated one
+	// to the operator namespace
+	configMapInformer := informers.InformersFor(util.OpenShiftConfigNamespace)
+	c := &ConfigSyncController{
+		operatorClient:       operatorClient,
+		kubeClient:           kubeClient,
+		configMapLister:      configMapInformer.Core().V1().ConfigMaps().Lister(),
+		infrastructureLister: configInformers.Config().V1().Infrastructures().Lister(),
+		eventRecorder:        eventRecorder.WithComponentSuffix("ConfigSync"),
+	}
+	return factory.New().WithSync(c.sync).ResyncEvery(resyncInterval).WithSyncDegradedOnError(operatorClient).WithInformers(
+		operatorClient.Informer(),
+		configMapInformer.Core().V1().ConfigMaps().Informer(),
+	).ToController("ConfigSync", eventRecorder)
+}
+
+func (c *ConfigSyncController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorv1.Managed {
+		return nil
+	}
+
+	isMultiAZDeployment, err := isMultiAZDeployment()
+	if err != nil {
+		return err
+	}
+
+	infra, err := c.infrastructureLister.Get(infrastructureResourceName)
+	if err != nil {
+		return err
+	}
+
+	sourceConfig, err := c.configMapLister.ConfigMaps(util.OpenShiftConfigNamespace).Get(infra.Spec.CloudConfig.Name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// TODO: report error after some while?
+			klog.V(2).Infof("Waiting for config map %s from %s", infra.Spec.CloudConfig.Name, util.OpenShiftConfigNamespace)
+			return nil
+		}
+		return err
+	}
+
+	targetConfig, err := translateConfigMap(sourceConfig, isMultiAZDeployment)
+	if err != nil {
+		return err
+	}
+
+	_, _, err = resourceapply.ApplyConfigMap(ctx, c.kubeClient.CoreV1(), c.eventRecorder, targetConfig)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func isMultiAZDeployment() (bool, error) {
+	ci, err := getCloudInfo()
+	if err != nil {
+		return false, fmt.Errorf("couldn't collect info about cloud availability zones: %w", err)
+	}
+	klog.V(2).Infof("found cloud info: %+v", ci)
+	klog.V(2).Infof("found compute zones: %+v", ci.ComputeZones)
+	klog.V(2).Infof("found volume zones: %+v", ci.ComputeZones)
+	// We consider a cloud multiaz when it either have several different zones
+	// or compute and volumes are different.
+	return len(ci.ComputeZones) != 1 || len(ci.VolumeZones) != 1 || ci.ComputeZones[0] != ci.VolumeZones[0], nil
```

Comment:
We still have an issue where getCloudInfo() always return empty compute and volume zones. I believe this comes from the implementation of getCloudInfo() where we returned a cached value if it's not nil.

The first invocation triggers an error (perhaps because it's too early?) and we never fetch the cloud info again. The error is different based on the cloud where the CI runs.

On vexxhost-managed, [we get](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_openstack-cinder-csi-driver-operator/78/pull-ci-openshift-openstack-cinder-csi-driver-operator-master-e2e-openstack/1521544675687665664/artifacts/e2e-openstack/gather-extra/artifacts/pods/openshift-cluster-csi-drivers_openstack-cinder-csi-driver-operator-587d6cfc4f-mjrwv_openstack-cinder-csi-driver-operator.log):
```
E0503 18:17:23.602278       1 base_controller.go:272] ConfigSync reconciliation failed: couldn't collect info about cloud availability zones: failed to create a compute client: Post "https://rdo.vexxhost.ca:5000/v3/auth/tokens": dial tcp: i/o timeout
```

While on vexxhost-mecha (using self-signed cert) [we get](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_openstack-cinder-csi-driver-operator/78/pull-ci-openshift-openstack-cinder-csi-driver-operator-master-e2e-openstack-csi/1521544675729608704/artifacts/e2e-openstack-csi/gather-extra/artifacts/pods/openshift-cluster-csi-drivers_openstack-cinder-csi-driver-operator-587d6cfc4f-vqdzq_openstack-cinder-csi-driver-operator.log):
```
E0503 17:57:33.893921       1 base_controller.go:272] ConfigSync reconciliation failed: couldn't collect info about cloud availability zones: failed to create a compute client: Error parsing CA Cert from /etc/kubernetes/static-pod-resources/configmaps/cloud-config/ca-bundle.pem
```

Reply:
I've modified `getCloudInfo` so that it no longer caches the `CloudInfo` result internally. Instead, `isMultiAZDeployment` will cache the result returned by `getCloudInfo` but *only if* there are one or more volume and compute AZs. This makes sense since both Nova and Cinder have a default AZ of `nova` (which you shouldn't actually use! See warning [here](https://docs.openstack.org/nova/latest/admin/availability-zones.html)). Any request that doesn't return at least this is errant.

We probably want to discuss whether we even want to do this, since a user could configure it themselves and it feels a bit "magical"...